src/frontend/js/hooks/useResources/index.tsx

```diff
@@ -0,0 +1,69 @@
+import type { QueryKey, UseQueryOptions } from '@tanstack/react-query';
+import { MessageDescriptor } from 'react-intl';
+import { HttpError } from '../../utils/errors/HttpError';
+import { ApiResourceInterface } from '../../types/Joanie';
+import { useResourcesOmniscient } from './useResourcesOmniscient';
+import { useResourcesRoot } from './useResourcesRoot';
+
+export interface ResourcesQuery {
+  id?: string;
+}
+
+export interface Resource {
+  id: string;
+}
+
+export type QueryOptions<TData extends Resource> = Omit<
+  UseQueryOptions<unknown, HttpError, TData[]>,
+  'queryKey' | 'queryFn'
+>;
+
+export interface UseResourcesProps<
+  TData extends Resource,
+  TResourceQuery extends ResourcesQuery = ResourcesQuery,
+  TApiResource extends ApiResourceInterface<TData> = ApiResourceInterface<TData>,
+> {
+  queryKey: QueryKey;
+  filters?: TResourceQuery;
+  apiInterface: () => TApiResource;
+  omniscient?: boolean;
+  session?: boolean;
+  localized?: boolean;
+  messages?: Record<string, MessageDescriptor>;
+  queryOptions?: QueryOptions<TData>;
+  frozenQueryKey?: boolean;
+}
+
+export const useResourcesCustom = <
+  TData extends Resource,
+  TResourceQuery extends ResourcesQuery = ResourcesQuery,
+  TApiResource extends ApiResourceInterface<TData> = ApiResourceInterface<TData>,
+>(
+  props: UseResourcesProps<TData, TResourceQuery, TApiResource>,
+) => {
+  if (props.omniscient) {
+    return useResourcesOmniscient(props);
+  }
+  return useResourcesRoot(props);
+};
+
+export const useResources =
+  <
+    TData extends Resource,
+    TResourceQuery extends ResourcesQuery = ResourcesQuery,
+    TApiResource extends ApiResourceInterface<TData> = ApiResourceInterface<TData>,
+  >(
+    props: UseResourcesProps<TData, TResourceQuery, TApiResource>,
+  ) =>
+  (filters?: TResourceQuery, queryOptions?: QueryOptions<TData>) => {
+    return useResourcesCustom({ ...props, filters, queryOptions });
+  };
+
+export const useResource =
+  <TData extends Resource>(props: UseResourcesProps<TData>) =>
+  (id?: string, queryOptions?: QueryOptions<TData>) => {
+    const resources = useResources<TData>(props);
+    const res = resources({ id }, { ...queryOptions, enabled: !!id });
```

Comment:
I think we should be able to provide filters here.

```suggestion
  (id?: string, filters: Omit<TResourceQuery, 'id'> = {}, queryOptions?: QueryOptions<TData>) => {
    const resources = useResources<TData>(props);
    const res = resources({ id, ...filters }, { ...queryOptions, enabled: !!id });
```

Reply:
I'm adding it, but what's the point than adding filters on top of an `id` ? 🤔

Reply:
In case of product we would like to ensure that the product we are requesting is well linked to the current course.
`https://api.com/v1.0/api/products/<product_uid>?course=00001`